### ISSUE TITLE
[Bug]: chunk_bwd_dv_local 在特定 GQA head 配置下触发 NPU kernel hang

### ISSUE THREAD

**Issue by SEU-ZJY** [flashserve/flash-linear-attention-npu#462]
### 环境信息

python version: 3.9.25
CANN version: version=26.1.0
CANN public version: 9.2.0
torch version: 2.7.1+cpu
torch_npu version: 2.7.1.post6.dev20260723
triton-ascend version: 3.2.0
ascend：910B

### 最小复现

```python
import math
import torch
from fla_npu.ops import ascendc as ascendc_ops

torch.npu.config.allow_internal_format = False
torch.npu.set_compile_mode(jit_compile=False)
torch.npu.set_device(0)

B, HK, HV, T, K, V, chunk_size, scale = 1, 15, 120, 1, 128, 128, 64, 1.0 / math.sqrt(K)

torch.manual_seed(1080)
q = (torch.rand(B, HK, T, K) * 2 - 1).to(torch.float16).npu()
k = (torch.rand(B, HK, T, K) * 2 - 1).to(torch.float16).npu()
d_o = (torch.rand(B, HV, T, V) * 2 - 1).to(torch.float16).npu()
g = (-torch.rand(B, HV, T)).to(torch.float32).npu()

print("NPU BEFORE OP", flush=True)
dv = ascendc_ops.npu_chunk_bwd_dv_local(
    q,
    k,
    d_o,
    g,
    scale=scale,
    chunk_size=chunk_size,
    g_gamma=None,
    A=None,
    cu_seqlens=None,
    chunk_indices=None,
)
print("NPU AFTER OP", flush=True)

torch.npu.synchronize()
print("NPU AFTER SYNC", flush=True)


复现参数：


q:             fp16 [1, 15, 1, 128]
k:             fp16 [1, 15, 1, 128]
d_o:           fp16 [1, 120, 1, 128]
g:             fp32 [1, 120, 1]
scale:         1 / sqrt(128)
chunk_size:    64
g_gamma:       None
A:             None
cu_seqlens:    None
chunk_indices: None
```

### 完整报错

```text
单独运行上述 case 时没有立即产生 Python exception，而是稳定卡在：

torch.npu.synchronize()

日志表现为：

NPU BEFORE OP
NPU AFTER OP

但 `torch.npu.synchronize()` 无法返回，因此不会打印：

NPU AFTER SYNC
```

### 期望行为

根据当前接口约束，`HV % HK == 0` 的 GQA head 配置应能够正常执行。

如果该 head 配置实际上不在算子支持范围内，也期望在参数校验或 tiling 阶段明确返回错误，而不是 kernel launch 后一直无法完成，导致 NPU stream 和后续测试任务被阻塞。

### 其他补充

对 `HK`、`HV` 进行了单变量对照测试，结果如下：

| HK | HV  | HV / HK | 结果   |
| -- | --- | ------- | ---- |
| 2  | 16  | 8       | PASS |
| 14 | 112 | 8       | PASS |
| 15 | 120 | 8       | HANG |
| 16 | 128 | 8       | HANG |
| 15 | 15  | 1       | PASS |
| 15 | 30  | 2       | PASS |
| 15 | 60  | 4       | PASS |
| 30 | 120 | 4       | PASS |

**Comment by SEU-ZJY** [comment]
另外又遇到了一个可以稳定复现阻塞的用例：

```python
import torch
import torch_npu
from fla_npu.ops import ascendc

torch.npu.set_device(0)

device = "npu:0"

q = torch.empty(
    (1, 17, 64, 128),
    dtype=torch.float16,
    device=device,
).uniform_(-1.0, 1.0)

k = torch.empty(
    (1, 17, 64, 128),
    dtype=torch.float16,
    device=device,
).uniform_(-1.0, 1.0)

do = torch.empty(
    (1, 119, 64, 128),
    dtype=torch.float16,
    device=device,
).uniform_(-1.0, 1.0)

g = torch.empty(
    (1, 119, 64),
    dtype=torch.float16,
    device=device,
).uniform_(-1.0, 0.0)

scale = 0.08838834764831843
chunk_size = 64

print("before op", flush=True)

out = ascendc.chunk_bwd_dv_local(
    q,
    k,
    do,
    g,
    scale,
    chunk_size,
)

print("after op, before sync", flush=True)

torch.npu.synchronize()

print("after sync", flush=True)
```

执行：

```bash
timeout 300s python repro_2226.py
```

输出：

```text
before op
[W904 15:00:27.010556421 TensorFactories.cpp:340] Warning: Cannot create tensor with interal format while allow_internel_format=False, tensor will be created with base format. (function operator())
after op, before sync
```

程序不会继续打印：

```text
after sync
```

等待 `timeout 300s` 终止程序后执行：

```bash
echo $?
```

输出：

```text
124
```

说明程序是由于执行超过 300 秒被 `timeout` 强制终止。

该用例对应的参数为：

```text
q:             fp16 [1, 17, 64, 128]
k:             fp16 [1, 17, 64, 128]
d_o:           fp16 [1, 119, 64, 128]
g:             fp16 [1, 119, 64]
scale:         0.08838834764831843
chunk_size:    64
HK:            17
HV:            119
HV / HK:       7
```

其中同样满足：

```text
HV % HK == 0
```

但算子调用返回后，NPU stream 在 `torch.npu.synchronize()` 阶段持续阻塞，无法正常完成。

**Comment by chen-linxin** [comment]
问题原因已经定位：QK workspace 只有 3 个 ring slot，但 QK ready/free 与 gated ready/free 原先都采用每 15 次批量反向确认。每个 QK head 会派生 `hRatio` 个 gated head，两条同步链推进速度不同。

在 `H_qk=15`、`H_do=120`、`hRatio=8` 的配置中，AIC 发送第 15 个 `QK_READY` 后等待 QK 批量确认；AIV 则在消费该 ready 前到达第 105 个 `GATED_READY`，开始等待 gated 批量确认。AIC 无法继续消费 gated 事件，AIV也无法继续消费 QK ready，最终形成循环等待。该问题位于 $S_h=K_hQ_h^T$、gating 以及 $dV_j=P_jdO_j$ 三个计算阶段之间的核间同步边界，并非数值公式错误。

本次修改：

- 将 QK ready/free 改为逐 QK head 的 `CrossCoreFlag` 信用同步。
- AIV 将 QK 数据搬入 UB 队列后立即归还 QK free，解除 QK ready 的计数压力。
- gated ready/free 保持不变，仍保证全部 `hRatio` 个 gated workspace 写回后才复用对应 ring slot。
- 普通 Vector 和 Ascend 950 arch35 实现均已同步修改。
- 增加 Issue 原始配置的 Fast-Kernel 回归测试，并补充 README 同步说明。

验证状态：

- A2 Fast-Kernel 测试 14/14 通过。
- Issue 原始配置通过，最大绝对误差 `0.001953`。
- `H_qk=17`、`H_do=119`、`hRatio=7` 补充配置通过，最大绝对误差 `0.007780`。
- A2 Example ST 前向及全部梯度精度检查通过。
- sanitizer synccheck 未发现 ERROR、死锁或超时；racecheck 报告 `No error detected`。
- A2、A3、A5 单算子包编译通过；A3/A5 本轮仅完成交叉编译。
- 未观察到性能回退。

修复 PR：https://github.com/flashserve/flash-linear-attention-npu/pull/465

**Comment by chen-linxin** [comment]
补充处理了 ATK 用例中 `g_dtype` 和变长元数据未被 executor 实际消费的覆盖缺口。

修改内容：

- 精度用例扩展至 200 条，覆盖 BF16/FP16、`g` 的 BF16/FP16/FP32、`V=128/256`、`chunk_size=64/128`、定长及变长接口路径。
- executor 现已根据 `case_spec` 实际构造 `g_dtype` 和 `cu_seqlens/chunk_indices`，CPU 标杆与 NPU DUT 使用相同场景。
- 三份 JSON 全部统一为 `mixed_tolerance_bm`。

验证状态：

- 精度 200/200 通过。
- 确定性 16/16 通过，每条循环 50 次。
- 内存检测 16/16 通过，目标算子区间非法访问为 0。
- 性能测试 8/8 完成，平均耗时 `4.438 us`。

修复 PR：https://github.com/flashserve/flash-linear-attention-npu/pull/465

**Comment by chen-linxin** [comment]
补充按仓库最新精度测试标准完成重新验证。

本次调整：

- 每个独立精度场景使用 3 个固定 seed，并运行全部 `(case, seed)` 组合。
- 200 个基础场景扩展为 600 个三 seed 组合，另增加 3 个 `H_qk=15`、`H_do=120`、`hRatio=8` 的 Issue 定向回归组合，共 603 个精度用例。
- 全部用例采用 `mixed_tolerance_bm`，使用 CPU FP64 标杆和 GM 初始化。
- 测试 JSON 与 README 已同步更新。

验证状态：

- A2、ATK 26.8.8：603/603 执行成功，执行失败 0。
- 精度 603/603 通过，通过率 100%，最终结果 `Pass`。
- `hRatio=8` 定向回归 3/3 通过，未出现 kernel hang。

修复 PR：https://github.com/flashserve/flash-linear-attention-npu/pull/465

**Comment by chen-linxin** [comment]
v26.9.0 回移植 PR 已提交：#473

问题原因：
QK workspace 只有 3 个 ring slot，旧实现的 QK ready/free 与 gated ready/free 都采用每 15 次批量反向确认。高 `hRatio` 下两条同步链推进速度不同，AIC 会在第 15 个 QK ready 后等待 QK free，而 AIV 可能先到 gated 确认边界等待 gated free，形成循环等待。

修改内容：
- QK ready/free 改为逐 QK head 的 `CrossCoreFlag` 信用同步。
- AIV 将 QK 搬入 UB 后，立即由两个 Vector 子核以 Mode 2 共同返回一份 QK free。
- gated ready/free 保持不变，继续保证 workspace 安全复用。
- 同步修改普通与 arch35 Vector 实现，并补充算子 README。

验证状态：
- A2：wheel 构建、安装和 API 检查通过；ATK 混合容差精度 603/603、确定性 16/16（每例 50 次）、内存检测 16/16、性能 8/8 完成；Issue 原始配置的 3 个 seed 均通过且无 hang，整体性能较基线提升 1.82%。
- A3：单算子包交叉编译通过。
- A5：wheel 构建、安装和 API 检查通过；Issue 原始配置 3/3 通过且无 hang；性能 8/8 完成，三轮整体中位波动 +2.11%。完整 ATK 矩阵受当前 GM 初始化资源占用限制，已在 PR 中如实记录。

PR：https://github.com/flashserve/flash-linear-attention-npu/pull/473
线上 full NPU CI 补充：
- 已触发 [full NPU CI](https://github.com/flashserve/flash-linear-attention-npu/actions/runs/34029586754)，A2/A3/A5 整包与单算子编译 6/6 通过。
- full 流程旧单测 7/11 通过，`chunk_bwd_dv_local` 在 13 秒内返回失败但未 hang，同时有 3 个未修改算子失败；该旧测试使用未按当前规范校准的 gate 值域，与最新 ATK 混合容差 603/603 结果不一致。
- 流程随后因自定义 OPP `op_api` 安装路径未被主干可信 CI 识别而中止，Example ST 无法解析算子符号，精度报告未生成。因此线上门禁当前仍为失败，不能宣称 full CI 通过；详细限制已写入 PR。
- 本回移植 PR 仍只包含算子同步修复。上述 v26.9 旧验证流程与主干可信 CI 的兼容问题未混入本 PR。
